File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/IntradayBarMessage.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/IntradayBarMessage.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Service;
import com.bloomberglp.blpapi.Message;
import com.bloomberglp.blpapi.CorrelationID;
import java.lang.StringBuilder;
import java.util.List;

public class IntradayBarMessage extends Message
{
	private final IntradayBarElementData _parent;
    private final IntradayBarElementResponseError _responseError;
    private final boolean _isResponseError;
	
    public IntradayBarMessage(CorrelationID corr, Service service, String security) throws Exception
    {
    	super(new Name("IntradayBarResponse"), corr, service);
        this._responseError = new IntradayBarElementResponseError(security);
        this._parent = null;
        this._isResponseError = true;
    }
    
    
	public IntradayBarMessage(CorrelationID corr, List<IntradayBarTickDataType> bars, Service service) throws Exception
	{
		super(new Name("IntradayBarResponse"), corr, service);
		this._parent = new IntradayBarElementData(bars);
        this._responseError = null;
        this._isResponseError = false;
	}
	
	public boolean hasElement(String name) throws Exception
	{
		return this.hasElement(name, false);
	}
	
	public boolean hasElement(String name, boolean excludeNullElements) throws Exception
	{
		return 
				(this._isResponseError && name.equals(this._responseError.name().toString())) ||
				(!this._isResponseError && name.equals(this._parent.name().toString()));
	}
	
	public Element getElement(String name) throws Exception
	{
		if(this._isResponseError)
		{
			if(name.equals(this._responseError.name().toString()))
				return this._responseError;
		}
		else
		{
			if(name.equals(this._parent.name().toString()))
				return this._parent;
		}
		
		throw new Exception("not implemented. names are case-sensitive.");
	}
	
	public String topicName()
	{
		return "";
	}
	
	public int numElements()
	{
		return 1;
	}
	
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append("IntradayBarResponse (choice) = {" + System.getProperty("line.separator"));
        
        try
    	{
	        if(this._isResponseError)
				result.append(this._responseError.prettyPrint(1));
			else        
	        	result.append(this._parent.prettyPrint(1));
    	}
        catch (Exception e) { }
        
        result.append("}");

        return result.toString();
    }
	
}

File: BEmu_cpp/src/IntradayBarRequest/IntradayBarMessage.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayBarRequest/IntradayBarMessage.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "BloombergTypes/MessagePtr.h"
#include "IntradayBarRequest/IntradayBarMessage.h"
#include "IntradayBarRequest/IntradayBarElementData.h"
#include "IntradayBarRequest/IntradayBarMessage.h"
#include "IntradayBarRequest/IntradayBarElementResponseError.h"
#include "BloombergTypes/Name.h"

namespace BEmu
{
	namespace IntradayBarRequest
	{
		IntradayBarMessage::IntradayBarMessage(const CorrelationId& corr, const Service& service, const std::string& security) : 
			MessagePtr(Name("IntradayBarResponse"), corr),
			_responseError(new IntradayBarElementResponseError(security)),
			_parent()
		{
            this->_isResponseError = true;
		}

		IntradayBarMessage::IntradayBarMessage(const CorrelationId& corr, const Service& service, const std::vector< boost::shared_ptr<IntradayBarTickDataType> >& bars) :
			MessagePtr(Name("IntradayBarResponse"), corr),
			_responseError(),
			_parent(new IntradayBarElementData(bars))
		{
            this->_isResponseError = false;
		}

		IntradayBarMessage::~IntradayBarMessage()
		{
		}

		const char* IntradayBarMessage::topicName() const
		{
			return "";
		}

		size_t IntradayBarMessage::numElements() const
		{
			return 1;
		}

		boost::shared_ptr<ElementPtr> IntradayBarMessage::getElement(const char* name) const
		{
            if (this->_isResponseError)
			{
				if(strncmp(name, "responseError", 14) == 0)
					return boost::dynamic_pointer_cast<ElementPtr>(this->_responseError);
				else
					throw messageEx;
			}
			else
			{
				if(strncmp(name, "barData", 8) == 0)
					return boost::dynamic_pointer_cast<ElementPtr>(this->_parent);
				else
					throw messageEx;
			}
		}

		bool IntradayBarMessage::hasElement(const char* name, bool excludeNullElements) const
		{
			return
				(strncmp(name, "responseError", 14) == 0 && this->_isResponseError) ||
				(strncmp(name, "barData", 8) == 0 && !this->_isResponseError);
		}

		std::ostream& IntradayBarMessage::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			stream << "IntradayBarResponse (choice) = {" << std::endl;

			if(this->_isResponseError)
				this->_responseError->print(stream, level, spacesPerLevel);
			else
				this->_parent->print(stream, level, spacesPerLevel);

			stream << '}' << std::endl;
			return stream;
		}

	}
}